File: day21/part2.rb
def addr(registers, instruction)
  _, a, b, c = instruction
  registers[c] = registers[a] + registers[b]
  end

def addi(registers, instruction)
  _, a, b, c = instruction
  registers[c] = registers[a] + b
end

def mulr(registers, instruction)
  _, a, b, c = instruction
  registers[c] = registers[a] * registers[b]
end

def muli(registers, instruction)
  _, a, b, c = instruction
  registers[c] = registers[a] * b
end

def banr(registers, instruction)
  _, a, b, c = instruction
  registers[c] = registers[a] & registers[b]
end

def bani(registers, instruction)
  _, a, b, c = instruction
  registers[c] = registers[a] & b
end

def borr(registers, instruction)
  _, a, b, c = instruction
  registers[c] = registers[a] | registers[b]
end

def bori(registers, instruction)
  _, a, b, c = instruction
  registers[c] = registers[a] | b
end

def setr(registers, instruction)
  _, a, _, c = instruction
  registers[c] = registers[a]
end

def seti(registers, instruction)
  _, a, _, c = instruction
  registers[c] = a
end

def gtir(registers, instruction)
  _, a, b, c = instruction
  registers[c] = a > registers[b] ? 1 : 0
end

def gtri(registers, instruction)
  _, a, b, c = instruction
  registers[c] = registers[a] > b ? 1 : 0
end

def gtrr(registers, instruction)
  _, a, b, c = instruction
  registers[c] = registers[a] > registers[b] ? 1 : 0
end

def eqir(registers, instruction)
  _, a, b, c = instruction
  registers[c] = a == registers[b] ? 1 : 0
end

def eqri(registers, instruction)
  _, a, b, c = instruction
  registers[c] = registers[a] == b ? 1 : 0
end

def eqrr(registers, instruction)
  _, a, b, c = instruction
  registers[c] = registers[a] == registers[b] ? 1 : 0
end

lines = File.readlines('input')
lines.map!(&:split)
operations = %w[bani banr muli setr bori eqrr gtir mulr gtrr seti gtri eqri addi borr eqir addr]
opcodes = operations.map.with_index(1).to_h
registers = [0, 0, 0, 0, 0, 0]
ip_register = lines.shift.last.to_i
ip = 0

seen = Hash.new(false)
last = 0
while line = lines[ip]
  operation = line[0]
  instruction = line.map(&:to_i)
  registers[ip_register] = ip
  # print "ip=#{ip} #{registers} #{line.join(' ')} "
  if ip == 28
    target = registers[instruction[1]]
    break if seen[target]

    seen[target] = true
    last = target
  end
  Object.send(operation, registers, instruction)
  ip = registers[ip_register]
  ip += 1
  # puts registers.to_s
end

puts last

File: day21/part2.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"io/ioutil"
)

func main() {
	file, _ := ioutil.ReadFile("input")
	lines := strings.Split(strings.TrimSpace(string(file)), "\n")

	registers := [...]int{0, 0, 0, 0, 0, 0}
	ipRegister, _ := strconv.Atoi(strings.Split(lines[0], " ")[1])
	ip := 0

	seen := map[int]bool{}
	last := 0
	lines = lines[1:]
	for {
		if ip>=len(lines) {
			break
		}
		registers[ipRegister] = ip

		instruction := strings.Split(lines[ip], " ")

		operation := instruction[0]
		a, _ := strconv.Atoi(instruction[1])
		b, _ := strconv.Atoi(instruction[2])
		c, _ := strconv.Atoi(instruction[3])

		if (ip==28) {
			target := registers[a]
			if _, found :=seen[target]; found {
				break 
			}
			seen[target] = true
			last = target
		}

		switch operation {
		case "addr":
			registers[c] = registers[a] + registers[b]
		case "addi":
			registers[c] = registers[a] + b
		case "mulr":
			registers[c] = registers[a] * registers[b]
		case "muli":
			registers[c] = registers[a] * b
		case "banr":
			registers[c] = registers[a] & registers[b]
		case "bani":
			registers[c] = registers[a] & b
		case "borr":
			registers[c] = registers[a] | registers[b]
		case "bori":
			registers[c] = registers[a] | b
		case "setr":
			registers[c] = registers[a]
		case "seti":
			registers[c] = a
		case "gtir":
			if a > registers[b] {
				registers[c] = 1
			} else {
				registers[c] = 0
			}
		case "gtri":
			if registers[a] > b {
				registers[c] = 1
			} else {
				registers[c] = 0
			}
		case "gtrr":
			if registers[a] > registers[b] {
				registers[c] = 1
			} else {
				registers[c] = 0
			}
		case "eqir":
			if a == registers[b] {
				registers[c] = 1
			} else {
				registers[c] = 0
			}
		case "eqri":
			if registers[a] == b {
				registers[c] = 1
			} else {
				registers[c] = 0
			}
		case "eqrr":
			if registers[a] == registers[b] {
				registers[c] = 1
			} else {
				registers[c] = 0
			}
		default:
			panic("unrecognized operation")
		}	
		ip = registers[ipRegister]
  		ip++
	}
	fmt.Println(last)
}